frontend/src/lib/ReportNode.ts

```diff
@@ -0,0 +1,509 @@
+import { makeElementWithInfoText, notUndefined } from "./utils"
+
+const NO_REPORT_ID = "NO_REPORT_ID"
+
+/**
+ * An immutable node of the "Report Data Tree".
+ *
+ * Trees are composed of `ElementNode` leaves, which contain data about
+ * a single visual element, and `BlockNode` branches, which determine the
+ * layout of a group of children nodes.
+ *
+ * A simple tree might look like this:
+ *
+ *   ReportRoot
+ *   ├── BlockNode ("main")
+ *   │   ├── ElementNode (text: "Ahoy, Streamlit!")
+ *   │   └── ElementNode (button: "Don't Push This")
+ *   └── BlockNode ("sidebar")
+ *       └── ElementNode (checkbox: "Batten The Hatches")
+ *
+ * To build this tree, the frontend receives `Delta` messages from Python,
+ * each of which corresponds to a tree mutation ("add an element",
+ * "add a block", "add rows to an existing element"). The frontend builds the
+ * tree bit by bit in response to these `Delta`s.
+ *
+ * To render the app, the `ReportView` class walks this tree and outputs
+ * a corresponding DOM structure, using React, that's essentially a mapping
+ * of `ReportElement` -> `ReactNode`. This rendering happens "live" - that is,
+ * the app is re-rendered each time a new `Delta` is received.
+ *
+ * Because the app gets re-rendered frequently, updates need to be fast.
+ * Our React components - the building blocks of the app - are "pure"
+ * (see https://reactjs.org/docs/react-api.html#reactpurecomponent), which
+ * means that React uses shallow comparison to determine which ReactNodes to
+ * update.
+ *
+ * Thus, each node in our tree is _immutable_ - any change to a `ReportNode`
+ * actually results in a *new* `ReportNode` instance. This occurs recursively,
+ * so inserting a new `ElementNode` into the tree will also result in new
+ * `BlockNode`s for each of that Element's ancestors, all the way up to the
+ * root node. Then, when React re-renders the app, it will re-traverse the new
+ * nodes that have been created, and rebuild just the bits of the app that
+ * have changed.
+ */
+export interface ReportNode {
+  /**
+   * The ID of the report this node was generated in. When a report finishes
+   * running, the app prunes all stale nodes.
+   */
+  readonly reportId: string
+
+  /**
+   * Return the ReportNode for the given index path, or undefined if the path
+   * is invalid.
+   */
+  getIn(path: number[]): ReportNode | undefined
+
+  /**
+   * Return a copy of this node with a new element set at the given index
+   * path. Throws an error if the path is invalid.
+   */
+  setIn(path: number[], node: ReportNode, reportId: string): ReportNode
+
+  /**
+   * Recursively remove children nodes whose reportID is no longer current.
+   * If this node should no longer exist, return undefined.
+   */
+  clearStaleNodes(currentReportId: string): ReportNode | undefined
+
+  /**
+   * Recursively add all Elements contained in the tree to a set.
+   */
+  getElements(elementSet: Set<Element>): void
```

Comment:
I think there are a few potential things to do here, but the key is that this is a recursive function that builds a set by visiting every node in the tree. Returning a new set from each function would be tons of overhead - you'd be creating tons of temporary Sets that are being union'd together and thrown away, union'd together and thrown away, over and over again.

An alternate pattern would be to do something like:

```typescript
getElements(set?: Set<Element>): Set<Element> {
  // If an existing set was passed in, use that. Else create a new one.
  if (set == null) {
      set = new Set<Element>()
  }
  // do the things
  return set
}
```

I think "copyElementsToSet" is a slight misnomer because it's not actually copying them, but I'd be down with "addElementsToSet" - yay?

Reply:
Both proposals look great to me!

Reply:
Cool, I just changed this to the "pass an optional Set and return the output set" pattern